Training/interface/DataLoader_main.h

```diff
@@ -472,7 +472,7 @@ class DataLoader {
                                     std::make_index_sequence<nFeaturesTypes>{});
 
         auto fillGrid = [&](auto _feature_idx, float value) {
-          if(static_cast<int>(_feature_idx) < 0) return;
+          if((static_cast<int>(_feature_idx) < 0) || !(std::isfinite(value))) return;
```

Comment:
I think throwing an exception is a better way to handle non-finite values. At the level of training it is desirable to have control of the inputs, i.e. we should know the domain of each input (including when it is not valid) and treat this cases before calling `fillGrid`.

Reply:
Do we have many variables that can have unexpected NaN values? I hope that the amount of variables is rather limited and we can treat each one separately to not loose "physics sense" why variable is undefined. E.g. if there is no track details, it is expected that `dxy_error` is not defined etc. If we start blindly screening all the variables, we risking to mis some errors in the input dataset or misuse some inputs.

Indeed, there is an exception thrown in the later parts of the code, I wasn't sure if it is still enabled by default. So then we shouldn't have check here.

Reply:
Looking at it again (and testing successfully with the other Scaling config), it's only for `pfCand_gamma_vertex_dz`/`_tauFL`. But even "blinding" these two inputs, wouldn't that cause the same issues in the example you're giving? Would it just be better to leave this as a comment in the code, so that one may locally make changes to skip NANs in specific scenarios such as this, where it's known that there are just a few taus with NANs?

Reply:
The question is from where do you know it? If NaN screeening would be in place, would you have noticed that the variables in question sometimes ill-defined?
IMO, we should have a good knowledge of the variable domain which we are sure is up-to-date. For the future R&D, we should consider reduction of the "useless" or ill-defined input variables. E.g. it is very likely that the vertex information for pf gammas can be easily dropped without any performance losses.

Reply:
I believe it's in the feature scaling step, where I first encountered those NANs. The data are loaded differently in that step, right? So I thought that, yes, I would know beforehand.

Reply:
If I may add (quite a long) comment, not fully relevant to the ongoing discussion but rather with a purpose of broadening the perspective a bit and possibly stepping aside from our current workflow. So if you'd like we can discuss this later at some meeting.

I generally think that it is a good practise to understand the input data, otherwise one is prone to have "garbage in, garbage out" results. It is a broad question what it means "to understand the data", but if we firstly consider a simpler NaN problem, it looks to me that the approach "if any number is NaN -> mask it" brings less understanding of the input domain then "if training crashed because of some NaN  -> find NaNed feature and mask its values in DataLoader manually". It might be annoying, because reaching the convergence requires to go throw several crashed trainings, but it forces the user to have a look at the distributions of input features and therefore gain the understanding. That is why I tend to agree with Konstantin about avoiding "inclusive" masking and trying to be specific.

However, I've just realised that fundamentally, this approach is also careless in terms of understanding the data, because it only makes to think about the data when there is a problem. A more consistent approach would be for the user to spend some time investigating all the distributions prior to running the training. But this approach suffers from the curse of dimensionality since we have quite a lot of features to look through and that takes time. Also, it is hard to visualise a dataset if it doesn't fit into the memory.

Ideally, it would be nice to have some kind of dashboard which shows at once all the histograms for the specified dataset/input feature with all necessary statistics about it (and I bet that there is already existing solutions which do that). So the current compromise to automate this "understanding" procedure within our framework is the feature scaling step, at least that's what I was thinking of when I was working on it.
 
I had an idea that since we iterate over the _whole_ dataset anyway, why don't we collect some knowledge as we go, aside from the scaling parameters? One example would be [scanning for NaN/inf](https://github.com/cms-tau-pog/TauMLTools/blob/31d2a54717657453d80436ec8ef16504d428faa4/Training/python/feature_scaling.py#L144-L149), but then there is [`plot_quantile_ranges.py`](https://github.com/cms-tau-pog/TauMLTools/blob/master/Training/python/plot_quantile_ranges.py) which job is to plot the quantile ranges for each feature -> help to understand its domain. And also [`plot_scaling_convergence.py`](https://github.com/cms-tau-pog/TauMLTools/blob/master/Training/python/plot_scaling_convergence.py) which should show if there are outliers in the convergence procedure -> point to anomalous features. 

So in principle, after running the feature scaling step and fully studying its outputs one can proceed to the masking of the features  in DataLoader (either exclusively or inclusively) which it showed to be NaNed, whilst at the same time having more confidence in understanding the input data. There is the caveat though that currently feature scaling preprocessing isn't automatically in synch with DataLoader preprocessing, which would be of course good to have unified across the framework.